valueset.go

```diff
@@ -0,0 +1,315 @@
+	query := "MATCH (root:`NODESET` {nodeset_id: $id})-[:$id]->(m) return root,m"
+	idrec, err := tx.Run(query, map[string]interface{}{"id": nodesetId})
+	if err != nil {
+		return Nodeset{}, err
+	}
+	ns := Nodeset{
+		ID:     nodesetId,
+		Labels: make([]string, 0),
+		Data:   make(map[string]NodesetData),
+	}
+	first := false
+	for idrec.Next() {
+		rec := idrec.Record()
+		if !first {
+			nsCenter := rec.Values[0].(neo4j.Node)
+			ns.Labels = nsCenter.Labels
+			first = true
+		}
+		nsLeaf := rec.Values[1].(neo4j.Node)
+		var nsLeafEntityId string
+		for k := range nsLeaf.Props {
+			if k == ls.EntityIDTerm {
+				nsLeafEntityId = k
+			}
+		}
+		ns.Data[nsLeafEntityId] = NodesetData{
+			ID:         nsLeafEntityId,
+			Labels:     nsLeaf.Labels,
+			Properties: nsLeaf.Props,
+		}
+	}
+	return ns, nil
+}
+
+func Execute(tx neo4j.Transaction, oldNodeset, newNodeset *Nodeset, rootOp string, inserts, updates []NodesetData, deletions []string) error {
+	type groupedNodesetData struct {
+		labelExpr string
+		data      []NodesetData
+	}
+	groupByLabel := func(nodesetData []NodesetData) []groupedNodesetData {
+		ret := make([]groupedNodesetData, 0)
+		hm := make(map[string][]NodesetData)
+		for _, nsD := range nodesetData {
+			nsD.Labels = sort.StringSlice(nsD.Labels)
+			labelExpr := ""
+			hm[labelExpr] = append(hm[labelExpr], nsD)
+		}
+		for expr, nsData := range hm {
+			ret = append(ret, groupedNodesetData{labelExpr: expr, data: nsData})
+		}
+		return ret
+	}
+	type udata struct {
+		udata []map[string]interface{}
+	}
+	if len(updates) > 0 {
+		unwindData := make(map[string]udata)
+		for ix, update := range groupByLabel(updates) {
+			unwind := make([]map[string]interface{}, 0)
+			item := map[string]interface{}{
+				"props": update.data[ix].Properties,
+			}
+			unwind = append(unwind, item)
+			unwindData[update.labelExpr] = udata{udata: unwind}
+		}
+		for label, unwind := range unwindData {
+			query := fmt.Sprintf(`unwind $nodes as node MATCH (n) WHERE n.entityId = node.ID SET n=node.props, n%s`, label)
+			if _, err := tx.Run(query, map[string]interface{}{"nodes": unwind.udata}); err != nil {
+				return err
+			}
+		}
+	}
+	if len(inserts) > 0 {
+		unwindData := make(map[string]udata)
+		for ix, insert := range groupByLabel(inserts) {
+			unwind := make([]map[string]interface{}, 0)
+			item := map[string]interface{}{
+				"props": insert.data[ix].Properties,
+			}
+			unwind = append(unwind, item)
+			unwindData[insert.labelExpr] = udata{udata: unwind}
+		}
+		for label, unwind := range unwindData {
+			query := fmt.Sprintf(`unwind $nodes as node create (a%s) set a=node.props`, label)
+			if _, err := tx.Run(query, map[string]interface{}{"nodes": unwind.udata}); err != nil {
+				return err
+			}
+		}
+	}
+	if len(deletions) > 0 {
+		if _, err := tx.Run("MATCH (n) WHERE n.entityId in $deletes DETACH DELETE n", map[string]interface{}{"deletes": deletions}); err != nil {
```

Comment:
This will not use indexes. So we should collect NodesetData for deletions as well, and use groupings to delete by using labels